zcash_extras/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+    /// a chain reorg might invalidate some stored state, this method must be
+    /// implemented in order to allow users of this API to "reset" the data store
+    /// to correctly represent chainstate as of a specified block height.
+    ///
+    /// After calling this method, the block at the given height will be the
+    /// most recent block and all other operations will treat this block
+    /// as the chain tip for balance determination purposes.
+    ///
+    /// There may be restrictions on how far it is possible to rewind.
+    async fn rewind_to_height(&mut self, block_height: BlockHeight) -> Result<(), Self::Error>;
+}
+
+/// This trait provides a generalization over shielded output representations.
+pub trait ShieldedOutput {
+    fn index(&self) -> usize;
+    fn account(&self) -> AccountId;
+    fn to(&self) -> &PaymentAddress;
+    fn note(&self) -> &Note;
+    fn memo(&self) -> Option<&MemoBytes>;
+    fn is_change(&self) -> Option<bool>;
+    fn nullifier(&self) -> Option<Nullifier>;
+}
+
+impl ShieldedOutput for WalletShieldedOutput<Nullifier> {
+    fn index(&self) -> usize {
+        self.index
+    }
+    fn account(&self) -> AccountId {
+        self.account
+    }
+    fn to(&self) -> &PaymentAddress {
+        &self.to
+    }
+    fn note(&self) -> &Note {
+        &self.note
+    }
+    fn memo(&self) -> Option<&MemoBytes> {
+        None
+    }
+    fn is_change(&self) -> Option<bool> {
+        Some(self.is_change)
+    }
+
+    fn nullifier(&self) -> Option<Nullifier> {
+        Some(self.nf)
+    }
+}
+
+impl ShieldedOutput for DecryptedOutput {
+    fn index(&self) -> usize {
+        self.index
+    }
+    fn account(&self) -> AccountId {
+        self.account
+    }
+    fn to(&self) -> &PaymentAddress {
+        &self.to
+    }
+    fn note(&self) -> &Note {
+        &self.note
+    }
+    fn memo(&self) -> Option<&MemoBytes> {
+        Some(&self.memo)
+    }
+    fn is_change(&self) -> Option<bool> {
+        None
+    }
+    fn nullifier(&self) -> Option<Nullifier> {
+        None
+    }
+}
+
+/// A newtype wrapper for sqlite primary key values for the notes
+/// table.
+#[derive(Debug, Copy, Clone)]
+pub enum NoteId {
+    SentNoteId(i64),
+    ReceivedNoteId(i64),
+}
+
+impl fmt::Display for NoteId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            NoteId::SentNoteId(id) => write!(f, "Sent Note {}", id),
+            NoteId::ReceivedNoteId(id) => write!(f, "Received Note {}", id),
+        }
+    }
+}
+#[cfg(feature = "mainnet")]
+pub(crate) fn network() -> Network {
+    Network::MainNetwork
+}
+
+#[cfg(not(feature = "mainnet"))]
+pub(crate) fn network() -> Network {
+    Network::TestNetwork
+}
+
+#[cfg(feature = "mainnet")]
+pub(crate) fn sapling_activation_height() -> BlockHeight {
+    Network::MainNetwork
+        .activation_height(NetworkUpgrade::Sapling)
+        .unwrap()
+}
```

Comment:
```bash
cargo build -p zcash_extras --features mainnet
error: none of the selected packages contains these features: mainnet
```

There is no `mainnet` feature specified in `Cargo.toml`. So `network` will always return `TestNetwork`:
```rust
#[test]
fn check_network() {
    assert_eq!(network(), Network::TestNetwork);
}
```
```
cargo test -p zcash_extras
running 1 test
test check_network ... ok
```

Reply:
I'm removing the code as they are not used anyways. thanks

Reply:
modified